android/app/src/main/java/com/microsoft/codepush/common/managers/CodePushTelemetryManager.java

```diff
@@ -0,0 +1,234 @@
+     * @return new update report.
+     */
+    public CodePushDeploymentStatusReport buildUpdateReport(CodePushLocalPackage currentPackage) {
+        CodePushStatusReportIdentifier currentPackageIdentifier = buildPackageStatusReportIdentifier(currentPackage);
+        CodePushStatusReportIdentifier previousStatusReportIdentifier = getPreviousStatusReportIdentifier();
+        CodePushDeploymentStatusReport report = null;
+        if (currentPackageIdentifier != null) {
+            if (previousStatusReportIdentifier == null) {
+                removeStatusReportSavedForRetry();
+                report = new CodePushDeploymentStatusReport();
+                report.setLocalPackage(currentPackage);
+                report.setStatus(SUCCEEDED);
+            } else {
+
+                /* Compare identifiers as strings for simplicity */
+                if (!previousStatusReportIdentifier.toString().equals(currentPackageIdentifier.toString())) {
+                    removeStatusReportSavedForRetry();
+                    report = new CodePushDeploymentStatusReport();
+                    if (previousStatusReportIdentifier.hasDeploymentKey()) {
+                        String previousDeploymentKey = previousStatusReportIdentifier.getDeploymentKey();
+                        String previousLabel = previousStatusReportIdentifier.getVersionLabel();
+                        report.setLocalPackage(currentPackage);
+                        report.setStatus(SUCCEEDED);
+                        report.setPreviousDeploymentKey(previousDeploymentKey);
+                        report.setPreviousLabelOrAppVersion(previousLabel);
+                    } else {
+
+                        /* Previous status report was with a binary app version. */
+                        report.setLocalPackage(currentPackage);
+                        report.setStatus(SUCCEEDED);
+                        report.setPreviousLabelOrAppVersion(previousStatusReportIdentifier.getVersionLabel());
+                    }
+                }
+            }
+        }
+        return report;
+    }
+
+    /**
+     * Builds rollback report using current local package information.
+     *
+     * @param lastFailedPackage current local package information.
+     * @return new rollback report.
+     */
+    public CodePushDeploymentStatusReport buildRollbackReport(CodePushLocalPackage lastFailedPackage) {
+        CodePushDeploymentStatusReport report = new CodePushDeploymentStatusReport();
+        report.setLocalPackage(lastFailedPackage);
+        report.setStatus(FAILED);
+        return report;
+    }
+
+    /**
+     * Gets status report already saved for retry it's sending.
+     *
+     * @return report saved for retry sending.
+     * @throws JSONException if there was error of deserialization of report from json document.
+     */
+    public CodePushDeploymentStatusReport getStatusReportSavedForRetry() throws JSONException {
+        //TODO move to settings manager
```

Comment:
Add a blank line before comment?

Reply:
@sergey-akhalkov I'll fix it here and in other places, thx.